server/util/background/background.go

```diff
@@ -35,14 +72,23 @@ func ToBackground(ctx context.Context) context.Context {
 // other values stored in the context. For that, we have this beauty. Use it
 // to make a copy of your expired context and do your cleanup work.
 func ExtendContextForFinalization(parent context.Context, timeout time.Duration) (context.Context, context.CancelFunc) {
-	ctx := disconnectedContext{parent: parent}
-	// If the original context already had a deadline, ensure that the given timeout
-	// doesn't result in a new deadline that's even shorter.
-	if originalDeadline, ok := parent.Deadline(); ok {
-		remainingTime := time.Until(originalDeadline)
-		if remainingTime > timeout {
-			timeout = remainingTime
+	ctx := newDisconnectedContext(parent)
+	go func() {
+		// Wait for the parent context to be canceled, then after the grace
+		// period, cancel the extended context.
+		select {
+		case <-ctx.Done():
+			return // cancelled manually
+		case <-parent.Done():
 		}
-	}
-	return context.WithTimeout(ctx, timeout)
+		t := time.NewTimer(timeout)
+		defer timeutil.StopAndDrainTimer(t)
```

Comment:
I don't think you need this. Draining the timer only matters if you plan on reusing the timer.

Reply:
ah right, changed to just `t.Stop()`